rtc/pcf85363a.c

```diff
@@ -0,0 +1,371 @@
+			res = rtc_get_time(&rtc_time, RTC_TSR2_REG_ADDR, 1);
+		else if (oid->id == dev_id_last_powerfail)
+			res = rtc_get_time(&rtc_time, RTC_TSR3_REG_ADDR, 1);
+		else
+			return -EINVAL;
+
+		if (res < 0)
+			return res;
+
+
+		unix_time = rtc_time_to_unix(&rtc_time);
+		snprintf(buf, sizeof(buf), "%llu\n", unix_time);
+	}
+	else if (oid->id == dev_id_batt_powerfail) {
+		snprintf(buf, sizeof(buf), "%u\n", common.battery_failed);
+	}
+	else {
+		return -EINVAL;
+	}
+
+	available = (ssize_t)strlen(buf) - offs;
+	if (available <= 0)
+		return 0;
+	if (available < (ssize_t)len)
+		len = available;
+
+	memcpy(data, &buf[offs], len);
+	return len;
+}
+
+
+static void thread(void *arg)
+{
+	uint32_t port = (uint32_t)arg;
+	msg_t msg;
+	unsigned long int rid;
+
+	for (;;) {
+		if (msgRecv(port, &msg, &rid) < 0) {
+			memset(&msg, 0, sizeof(msg));
+			msgRespond(port, &msg, rid);
+			continue;
+		}
+
+		switch (msg.type) {
+			case mtOpen:
+			case mtClose:
+				msg.o.io.err = EOK;
+				break;
+			case mtWrite:
+				msg.o.io.err = -EINVAL;
+				break;
+			case mtRead:
+				msg.o.io.err = dev_read(&msg.i.io.oid, msg.i.io.offs, msg.o.size, msg.o.data);
+				break;
+			case mtDevCtl:
+				dev_ctl(&msg);
+				break;
+			default:
+				msg.o.io.err = -EINVAL;
+				break;
+		}
+
+		msgRespond(port, &msg, rid);
+	}
+}
+
+
+static int create_devs(uint32_t port)
+{
+	oid_t dev;
+
+	dev.port = port;
+	dev.id = dev_id_rtc;
+
+	if (create_dev(&dev, "rtc0") < 0) {
+		printf("rtc: could not create device rtc0\n");
+		return -1;
+	}
+
+	/* unix timestamp of the first time since last full bootup we've switched to battery supply */
+	dev.id = dev_id_first_powerfail;
+	if (create_dev(&dev, "rtc_first_powerfail") < 0) {
+		printf("rtc: could not create device rtc_first_powerfail\n");
+		return -1;
+	}
+
+	/* unix timestamp of the last time since we've switched to battery supply */
+	dev.id = dev_id_last_powerfail;
+	if (create_dev(&dev, "rtc_last_powerfail") < 0) {
+		printf("rtc: could not create device rtc_last_powerfail\n");
+		return -1;
+	}
+
+	/* 0/1 state if RTC battery supply failed (rtc contents is invalid) */
+	dev.id = dev_id_batt_powerfail;
+	if (create_dev(&dev, "rtc_batt_powerfail") < 0) {
+		printf("rtc: could not create device rtc_batt_powerfail\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+/* assume errors during setup are not fatal as they control periperial RTC functions, not the main one */
+static void setup_rtc(void)
+{
+	const uint8_t battery_check_magic = 0x5a;
+	const uint8_t event_config[] = {
+		RTC_TSR_MODE_REG_ADDR,
+		(0b10 << 6) | (0b001 << 2), /* TSR3 = LB, TSR2 = FB */
+	};
+	const uint8_t ram_config[] = {
+		RTC_RAM_BYTE_REG_ADDR,
+		battery_check_magic,
+	};
+	const uint8_t reset_flags[] = {
+		RTC_FLAGS_REG_ADDR,
+		0,
+	};
+
+	uint8_t buf[1] = { 0 };
+	int res;
+
+	if ((res = i2c_busWrite(DEV_ADDR, event_config, sizeof(event_config))) < 0)
+		printf("rtc: failed to setup the events configuration: %s\n", strerror(res));
```

Comment:
Maybe use `stderr` to output errors?

Reply:
written in detail above why I think it's better to use `stdout` (although I would prefer to use `syslog` directly because of multiple logging levels available).